### ISSUE TITLE
not compatible with io.js

### ISSUE THREAD

**Issue by adrai** [CrabDude/trycatch#41]
```
node -v && node setTimeout.js
v2.2.1
This is an asynchronous scoped error handler!
 Error: foo
    at [object Object].f (/Users/adrai/Projects/trycatch/examples/setTimeout.js:5:9)
    at Timer.<anonymous> (timers.js:89:15)
/Users/adrai/Projects/trycatch/lib/trycatch.js:245
    throw err
          ^
TypeError: Cannot read property 'length' of undefined
    at _stream_writable.js:352:12
    at _stream_writable.js:342:5
    at node.js:408:9
    at process._tickDomainCallback (node.js:378:13)
```

any idea why?

**Comment by trevnorris** [comment]
You'll need to provide the version of node and trycatch you're running. Also a script that can reproduce the problem. Preferably a minimal test case.

**Comment by adrai** [comment]
It's everything there ;-)
(master of trycatch) this test file: https://github.com/CrabDude/trycatch/blob/master/examples/setTimeout.js
node version: v2.2.1

**Comment by ChALkeR** [comment]
@trevnorris The version of `node` here is obviously `2.2.1`, and the file is [/examples/setTimeout.js](https://github.com/CrabDude/trycatch/blob/master/examples/setTimeout.js)

**Comment by trevnorris** [comment]
Ah. Thanks. Was reviewing from my phone and didn't catch the wordwrap'd path of the script.

**Comment by brendanashworth** [comment]
is there a version/versions of io.js/node that it *does* work on?

**Comment by adrai** [comment]
v1.7.1 works and v1.8.1 does NOT work anymore

**Comment by trevnorris** [comment]
What is the latest version of io.js/node that does work?

**Comment by adrai** [comment]
v1.7.1

**Comment by trevnorris** [comment]
Wow, okay. Sorry. That's the last time I do issue review from my phone while being distracted by something else.

**Comment by adrai** [comment]
No problem, I know you are a busy man ;-)

**Comment by adrai** [comment]
Any ideas/news?

**Comment by CrabDude** [comment]
I'll take a look at updating it this week.

**Comment by adrai** [comment]
Any news?

**Comment by brendanashworth** [comment]
@trevnorris looks to be up your alley: git bisect says it was introduced in https://github.com/nodejs/io.js/commit/10e31ba56c676bdcad39ccad22ea9117733b8eb5.

**Comment by trevnorris** [comment]
Does try/catch wrap any of node's internals?

**Comment by CrabDude** [comment]
Ha. That's pretty much all it does. More specifically, it wraps the
boundary layer APIs (e.g., fs.*, process.nextTick, and EventEmitter). A lot
of work has gone into making sure it works perfectly, but as is usually the
case with monkey patching, there are no guarantees. I'll take a look at
this tonight.

On Mon, Aug 3, 2015 at 1:11 PM, Trevor Norris <notifications@github.com>
wrote:

> Does try/catch wrap any of node's internals?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/CrabDude/trycatch/issues/41#issuecomment-127393467>.
>



-- 
Adam Crabtree
Noderiety
PO Box 3823
Los Altos, CA 94024
Telephone: *(*469) 777-8415 <https://mail.google.com/>
Email: dude@noderiety.com

This e-mail message is intended for the use of the addressee(s) only and
may contain information that is privileged-confidential and exempt from
disclosure under applicable law.  If the reader of the message is not the
intended recipient or agent responsible for delivering the message to the
intended recipient, you are hereby notified that any review, dissemination,
distribution or copying of this communication is strictly prohibited.  If
you receive this communication in error, please notify us immediately by
reply e-mail and delete the original message.  Thank you.

**Comment by trevnorris** [comment]
If it's overriding the global (process), then it will break the API internals depend on.

**Comment by trevnorris** [comment]
But should be a pretty simple fix.

**Comment by CrabDude** [comment]
Obviously I haven't taken a look at this yet. It's in the back of my mind. No update ATM. Will update here when I do.

**Comment by adrai** [comment]
Do you think we will get a fix before node v4.0 will be released?

**Comment by joshgillies** [comment]
Hey all. :) Any progress made here?

**Comment by CrabDude** [comment]
@adrai @joshgillies Tomorrow, **Wednesday, 8/26, I'm going to take care of this.** It's been a crazy 4 months, and I unfortunately haven't had time, but I need `trycatch` updated for a class I'm teaching, so I'll have this fixed ASAP now. Sorry for the inconvenience. PRs are of course welcome, but `trycatch` is a sticky tricky beast, so I appreciate that's a tall order.

**Comment by CrabDude** [comment]
Some more background on the delay...

`trycatch` has evolved to be more of a prevent application errors from requiring a restart by protecting the core stack, in addition to its original async try/catch and long-stack trace functionality. For a while now I've been needing/wanting to break those 3 components apart into separate modules: uncaughtApplicationException, async try/catch and long-stack-traces; and possibly still package them all in `trycatch`.

Since I've transitioned to teaching my classes with `async/await` first using babel.js, and thus using `Promise`, there's now the additional question of to what extent is `trycatch` necessary? (Not that it isn't, it's just obviously not AS necessary with synchronous `try/catch`, `async/await` and `Promise`) With the new version, I want to make sure it's compatible with `async/await`, `Promise`, babel and as a standalone. Still, I'll push a fix for the base functionality tomorrow before working on the breaking the components apart and adding `async/await` / `Promise` compatibility.

**Comment by adrai** [comment]
Thanks for the update :-)

**Comment by CrabDude** [comment]
So...... Stuff came up. Tomorrow's schedule is wide open, and this is the first item on my list.

**Comment by CrabDude** [comment]
Fixed @v1.5.12. It was really simple. Sad it took so long.

[`hookit` wasn't forwarding arguments for `process.nextTick`.](https://github.com/CrabDude/hookit/commit/1e98ae9f70265dbcced45550429667e396b7590d)

**Comment by CrabDude** [comment]
While I was at it, I also added support for strict mode, which was problematic since you can't use `stackFrame.getFunction()`, and it works in babel now.
